src/librustc_const_eval/pattern.rs

```diff
@@ -285,6 +286,39 @@ impl<'a, 'tcx> Pattern<'tcx> {
         debug!("Pattern::from_hir({:?}) = {:?}", pat, result);
         result
     }
+
+    // Returns true if the pattern cannot bind, as it would require a value of type `!` to have
+    // been constructed. This check is conservative.
+    pub fn is_unreachable(&self) -> bool {
```

Comment:
Similarly here, there is existing code that tries to make a similar judgement, I have to go digging around to find it, but I suspect we should be re-using it.

Reply:
I had a look around, and couldn't find anything that looked suitable, but I might have been looking in the wrong places. If you give me a pointer, I'll refactor it!

Reply:
Let me look around.